plugins/modem-manager/fu-mm-device.c

```diff
@@ -653,6 +728,86 @@ fu_mm_device_write_firmware_qmi_pdc (FuDevice *device, GBytes *fw, GArray **acti
 	return TRUE;
 }
 
+static gboolean
+fu_mm_device_mbim_open (FuMmDevice *self, GError **error)
+{
+	self->mbim_qdu_updater = fu_mbim_qdu_updater_new (self->port_mbim);
+	return fu_mbim_qdu_updater_open (self->mbim_qdu_updater, error);
+}
+
+static gboolean
+fu_mm_device_mbim_close (FuMmDevice *self, GError **error)
+{
+	g_autoptr(FuMbimQduUpdater) updater = NULL;
+
+	updater = g_steal_pointer (&self->mbim_qdu_updater);
+	return fu_mbim_qdu_updater_close (updater, error);
+}
+
+static gboolean
+fu_mm_device_write_firmware_mbim_qdu (FuDevice *device, GBytes *fw, GError **error)
+{
+	g_autoptr(FuArchive) archive = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	GBytes *data;
+	g_autoptr(XbBuilder) builder = xb_builder_new ();
+	g_autoptr(XbBuilderSource) source = xb_builder_source_new ();
+	g_autoptr(XbSilo) silo = NULL;
+	g_autoptr(GPtrArray) parts = NULL;
+	XbNode *part = NULL;
+	const gchar *filename = NULL;
+	const gchar *csum;
+	g_autofree gchar *csum_actual = NULL;
+	FuMmDevice *self = FU_MM_DEVICE (device);
+
+	/* decompress entire archive ahead of time */
+	archive = fu_archive_new (fw, FU_ARCHIVE_FLAG_IGNORE_PATH, error);
+	if (archive == NULL)
+		return FALSE;
+
+	locker = fu_device_locker_new_full (device,
+					    (FuDeviceLockerFunc) fu_mm_device_mbim_open,
+					    (FuDeviceLockerFunc) fu_mm_device_mbim_close,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* load the manifest of operations */
+	data = fu_archive_lookup_by_fn (archive, "flashfile.xml", error);
+	if (data == NULL)
+		return FALSE;
+	if (!xb_builder_source_load_bytes (source, data,
+					   XB_BUILDER_SOURCE_FLAG_NONE, error))
+		return FALSE;
+	xb_builder_import_source (builder, source);
+	silo = xb_builder_compile (builder, XB_BUILDER_COMPILE_FLAG_NONE, NULL, error);
+	if (silo == NULL)
+		return FALSE;
+
+	/* get all the operation parts */
+	parts = xb_silo_query (silo, "parts/part", 0, error);
+	if (parts == NULL)
+		return FALSE;
+	part = g_ptr_array_index (parts, 0);
```

Comment:
I think you can use `xb_silo_query_first() `here

Reply:
Yes, you are right, xb_silo_query_first() is more suitable here.